File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/listeners/DefaultInAppMessageViewLifecycleListener.java
package com.braze.ui.inappmessage.listeners;

import android.app.Activity;
import android.net.Uri;
import android.view.View;

import com.appboy.enums.Channel;
import com.braze.enums.inappmessage.ClickAction;
import com.braze.models.inappmessage.IInAppMessage;
import com.braze.models.inappmessage.IInAppMessageHtml;
import com.braze.models.inappmessage.IInAppMessageImmersive;
import com.braze.models.inappmessage.MessageButton;
import com.braze.support.BrazeFileUtils;
import com.braze.support.BrazeLogger;
import com.braze.support.BundleUtils;
import com.braze.support.WebContentUtils;
import com.braze.ui.BrazeDeeplinkHandler;
import com.braze.ui.actions.NewsfeedAction;
import com.braze.ui.actions.UriAction;
import com.braze.ui.inappmessage.BrazeInAppMessageManager;
import com.braze.ui.inappmessage.InAppMessageCloser;

import java.io.File;

public class DefaultInAppMessageViewLifecycleListener implements IInAppMessageViewLifecycleListener {
  private static final String TAG = BrazeLogger.getBrazeLogTag(DefaultInAppMessageViewLifecycleListener.class);

  @Override
  public void beforeOpened(View inAppMessageView, IInAppMessage inAppMessage) {
    // Note that the client method must be called before any default processing below
    getInAppMessageManager().getInAppMessageManagerListener().beforeInAppMessageViewOpened(inAppMessageView, inAppMessage);
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.beforeOpened called.");
    inAppMessage.logImpression();
  }

  @Override
  public void afterOpened(View inAppMessageView, IInAppMessage inAppMessage) {
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.afterOpened called.");

    // Note that the client method must be called after any default processing above
    getInAppMessageManager().getInAppMessageManagerListener().afterInAppMessageViewOpened(inAppMessageView, inAppMessage);
  }

  @Override
  public void beforeClosed(View inAppMessageView, IInAppMessage inAppMessage) {
    // Note that the client method must be called before any default processing below
    getInAppMessageManager().getInAppMessageManagerListener().beforeInAppMessageViewClosed(inAppMessageView, inAppMessage);
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.beforeClosed called.");
  }

  @Override
  public void afterClosed(IInAppMessage inAppMessage) {
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.afterClosed called.");
    getInAppMessageManager().resetAfterInAppMessageClose();
    if (inAppMessage instanceof IInAppMessageHtml) {
      startClearHtmlInAppMessageAssetsThread();
    }
    inAppMessage.onAfterClosed();

    // Note that the client method must be called after any default processing above
    getInAppMessageManager().getInAppMessageManagerListener().afterInAppMessageViewClosed(inAppMessage);
  }

  @Override
  public void onClicked(InAppMessageCloser inAppMessageCloser, View inAppMessageView, IInAppMessage inAppMessage) {
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.onClicked called.");
    inAppMessage.logClick();

    // Perform the in-app message clicked listener action from the host application first. This give
    // the app the option to override the values that are sent from the server and handle the
    // in-app message differently depending on where the user is in the app.
    //
    // To modify the default in-app message clicked behavior, mutate the necessary in-app message members. As
    // an example, if the in-app message were to navigate to the news feed when it was clicked, the
    // behavior can be cancelled by setting the click action to NONE.
    boolean handled = getInAppMessageManager().getInAppMessageManagerListener().onInAppMessageClicked(inAppMessage, inAppMessageCloser);

    if (!handled) {
      // Perform the default (or modified) in-app message clicked behavior.
      performInAppMessageClicked(inAppMessage, inAppMessageCloser);
    }
  }

  @Override
  public void onButtonClicked(InAppMessageCloser inAppMessageCloser, MessageButton messageButton, IInAppMessageImmersive inAppMessageImmersive) {
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.onButtonClicked called.");
    inAppMessageImmersive.logButtonClick(messageButton);

    boolean handled = getInAppMessageManager().getInAppMessageManagerListener().onInAppMessageButtonClicked(inAppMessageImmersive, messageButton, inAppMessageCloser);

    if (!handled) {
      // Perform the default (or modified) in-app message button clicked behavior.
      performInAppMessageButtonClicked(messageButton, inAppMessageImmersive, inAppMessageCloser);
    }
  }

  @Override
  public void onDismissed(View inAppMessageView, IInAppMessage inAppMessage) {
    BrazeLogger.d(TAG, "IInAppMessageViewLifecycleListener.onDismissed called.");
    getInAppMessageManager().getInAppMessageManagerListener().onInAppMessageDismissed(inAppMessage);
  }

  private void performInAppMessageButtonClicked(MessageButton messageButton, IInAppMessage inAppMessage, InAppMessageCloser inAppMessageCloser) {
    performClickAction(messageButton.getClickAction(), inAppMessage, inAppMessageCloser, messageButton.getUri(), messageButton.getOpenUriInWebview());
  }

  private void performInAppMessageClicked(IInAppMessage inAppMessage, InAppMessageCloser inAppMessageCloser) {
    performClickAction(inAppMessage.getClickAction(), inAppMessage, inAppMessageCloser, inAppMessage.getUri(), inAppMessage.getOpenUriInWebView());
  }

  private void performClickAction(ClickAction clickAction, IInAppMessage inAppMessage, InAppMessageCloser inAppMessageCloser, Uri clickUri, boolean openUriInWebview) {
    if (getInAppMessageManager().getActivity() == null) {
      BrazeLogger.w(TAG, "Can't perform click action because the cached activity is null.");
      return;
    }
    switch (clickAction) {
      case NEWS_FEED:
        inAppMessageCloser.close(false);
        NewsfeedAction newsfeedAction = new NewsfeedAction(BundleUtils.toBundle(inAppMessage.getExtras()),
            Channel.INAPP_MESSAGE);
        BrazeDeeplinkHandler.getInstance().gotoNewsFeed(getInAppMessageManager().getActivity(), newsfeedAction);
        break;
      case URI:
        inAppMessageCloser.close(false);
        UriAction uriAction = BrazeDeeplinkHandler.getInstance().createUriActionFromUri(clickUri, BundleUtils.toBundle(inAppMessage.getExtras()),
            openUriInWebview, Channel.INAPP_MESSAGE);
        BrazeDeeplinkHandler.getInstance().gotoUri(getInAppMessageManager().getActivity(), uriAction);
        break;
      case NONE:
        inAppMessageCloser.close(inAppMessage.getAnimateOut());
        break;
      default:
        inAppMessageCloser.close(false);
        break;
    }
  }

  private BrazeInAppMessageManager getInAppMessageManager() {
    return BrazeInAppMessageManager.getInstance();
  }

  private void startClearHtmlInAppMessageAssetsThread() {
    new Thread(() -> {
      Activity inAppMessageActivity = BrazeInAppMessageManager.getInstance().getActivity();
      if (inAppMessageActivity != null) {
        File internalStorageCacheDirectory = WebContentUtils.getHtmlInAppMessageAssetCacheDirectory(inAppMessageActivity);
        BrazeFileUtils.deleteFileOrDirectory(internalStorageCacheDirectory);
      }
    }).start();
  }
}

File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/listeners/DefaultInAppMessageViewLifecycleListener.kt
package com.braze.ui.inappmessage.listeners

import android.net.Uri
import android.view.View
import com.appboy.enums.Channel
import com.braze.coroutine.BrazeCoroutineScope
import com.braze.enums.inappmessage.ClickAction
import com.braze.models.inappmessage.IInAppMessage
import com.braze.models.inappmessage.IInAppMessageHtml
import com.braze.models.inappmessage.IInAppMessageImmersive
import com.braze.models.inappmessage.MessageButton
import com.braze.support.BrazeFunctionNotImplemented
import com.braze.support.BrazeLogger.Priority.W
import com.braze.support.BrazeLogger.brazelog
import com.braze.support.WebContentUtils.getHtmlInAppMessageAssetCacheDirectory
import com.braze.support.deleteFileOrDirectory
import com.braze.support.toBundle
import com.braze.ui.BrazeDeeplinkHandler
import com.braze.ui.actions.NewsfeedAction
import com.braze.ui.inappmessage.BrazeInAppMessageManager
import kotlinx.coroutines.launch

open class DefaultInAppMessageViewLifecycleListener : IInAppMessageViewLifecycleListener {
    private val inAppMessageManager: BrazeInAppMessageManager
        get() = BrazeInAppMessageManager.getInstance()

    override fun beforeOpened(inAppMessageView: View, inAppMessage: IInAppMessage) {
        // Note that the client method must be called before any default processing below
        inAppMessageManager.inAppMessageManagerListener.beforeInAppMessageViewOpened(inAppMessageView, inAppMessage)
        brazelog { "IInAppMessageViewLifecycleListener.beforeOpened called." }
        inAppMessage.logImpression()
    }

    override fun afterOpened(inAppMessageView: View, inAppMessage: IInAppMessage) {
        brazelog { "IInAppMessageViewLifecycleListener.afterOpened called." }

        // Note that the client method must be called after any default processing above
        inAppMessageManager.inAppMessageManagerListener.afterInAppMessageViewOpened(inAppMessageView, inAppMessage)
    }

    override fun beforeClosed(inAppMessageView: View, inAppMessage: IInAppMessage) {
        // Note that the client method must be called before any default processing below
        inAppMessageManager.inAppMessageManagerListener.beforeInAppMessageViewClosed(inAppMessageView, inAppMessage)
        brazelog { "IInAppMessageViewLifecycleListener.beforeClosed called." }
    }

    override fun afterClosed(inAppMessage: IInAppMessage) {
        brazelog { "IInAppMessageViewLifecycleListener.afterClosed called." }
        inAppMessageManager.resetAfterInAppMessageClose()
        if (inAppMessage is IInAppMessageHtml) {
            startClearHtmlInAppMessageAssetsThread()
        }
        inAppMessage.onAfterClosed()

        // Note that the client method must be called after any default processing above
        inAppMessageManager.inAppMessageManagerListener.afterInAppMessageViewClosed(inAppMessage)
    }

    @Suppress("DEPRECATION")
    override fun onClicked(
        inAppMessageCloser: com.braze.ui.inappmessage.InAppMessageCloser,
        inAppMessageView: View,
        inAppMessage: IInAppMessage
    ) {
        brazelog { "IInAppMessageViewLifecycleListener.onClicked called." }
        inAppMessage.logClick()

        // Perform the in-app message clicked listener action from the host application first. This give
        // the app the option to override the values that are sent from the server and handle the
        // in-app message differently depending on where the user is in the app.
        //
        // To modify the default in-app message clicked behavior, mutate the necessary in-app message members. As
        // an example, if the in-app message were to navigate to the news feed when it was clicked, the
        // behavior can be cancelled by setting the click action to NONE.
        @Suppress("SwallowedException")
        val wasHandled = try {
            val wasHandledLegacy = inAppMessageManager.inAppMessageManagerListener.onInAppMessageClicked(inAppMessage, inAppMessageCloser)
            brazelog { "Deprecated onInAppMessageClicked(inAppMessage, inAppMessageCloser) called." }
            wasHandledLegacy
        } catch (e: BrazeFunctionNotImplemented) {
            brazelog { "Using non-deprecated onInAppMessageClicked(inAppMessage)" }
            inAppMessageManager.inAppMessageManagerListener.onInAppMessageClicked(inAppMessage)
        }
        if (!wasHandled) {
            // Perform the default (or modified) in-app message clicked behavior.
            performInAppMessageClicked(inAppMessage, inAppMessageCloser)
        }
    }

    @Suppress("DEPRECATION")
    override fun onButtonClicked(
        inAppMessageCloser: com.braze.ui.inappmessage.InAppMessageCloser,
        messageButton: MessageButton,
        inAppMessageImmersive: IInAppMessageImmersive
    ) {
        brazelog { "IInAppMessageViewLifecycleListener.onButtonClicked called." }
        inAppMessageImmersive.logButtonClick(messageButton)
        @Suppress("SwallowedException")
        val wasHandled =
            try {
                inAppMessageManager.inAppMessageManagerListener.onInAppMessageButtonClicked(
                    inAppMessageImmersive,
                    messageButton,
                    inAppMessageCloser
                )
            } catch (e: BrazeFunctionNotImplemented) {
                inAppMessageManager.inAppMessageManagerListener.onInAppMessageButtonClicked(
                    inAppMessageImmersive,
                    messageButton
                )
            }
        if (!wasHandled) {
            // Perform the default (or modified) in-app message button clicked behavior.
            performInAppMessageButtonClicked(messageButton, inAppMessageImmersive, inAppMessageCloser)
        }
    }

    override fun onDismissed(inAppMessageView: View, inAppMessage: IInAppMessage) {
        brazelog { "IInAppMessageViewLifecycleListener.onDismissed called." }
        inAppMessageManager.inAppMessageManagerListener.onInAppMessageDismissed(inAppMessage)
    }

    @Suppress("DEPRECATION")
    private fun performInAppMessageButtonClicked(
        messageButton: MessageButton,
        inAppMessage: IInAppMessage,
        inAppMessageCloser: com.braze.ui.inappmessage.InAppMessageCloser
    ) {
        performClickAction(
            messageButton.clickAction,
            inAppMessage,
            inAppMessageCloser,
            messageButton.uri,
            messageButton.openUriInWebview
        )
    }

    @Suppress("DEPRECATION")
    private fun performInAppMessageClicked(inAppMessage: IInAppMessage, inAppMessageCloser: com.braze.ui.inappmessage.InAppMessageCloser) {
        performClickAction(
            inAppMessage.clickAction,
            inAppMessage,
            inAppMessageCloser,
            inAppMessage.uri,
            inAppMessage.openUriInWebView
        )
    }

    @Suppress("DEPRECATION")
    private fun performClickAction(
        clickAction: ClickAction,
        inAppMessage: IInAppMessage,
        inAppMessageCloser: com.braze.ui.inappmessage.InAppMessageCloser,
        clickUri: Uri?,
        openUriInWebview: Boolean
    ) {
        val activity = inAppMessageManager.activity
        if (activity == null) {
            brazelog(W) { "Can't perform click action because the cached activity is null." }
            return
        }
        when (clickAction) {
            ClickAction.NEWS_FEED -> {
                inAppMessageCloser.close(false)
                val newsfeedAction = NewsfeedAction(
                    inAppMessage.extras.toBundle(),
                    Channel.INAPP_MESSAGE
                )
                BrazeDeeplinkHandler.getInstance()
                    .gotoNewsFeed(activity, newsfeedAction)
            }
            ClickAction.URI -> {
                inAppMessageCloser.close(false)
                if (clickUri == null) {
                    brazelog { "clickUri is null, not performing click action" }
                    return
                }
                val uriAction = BrazeDeeplinkHandler.getInstance().createUriActionFromUri(
                    clickUri, inAppMessage.extras.toBundle(),
                    openUriInWebview, Channel.INAPP_MESSAGE
                )

                val appContext = inAppMessageManager.applicationContext
                if (appContext == null) {
                    brazelog { "appContext is null, not performing click action" }
                    return
                } else {
                    BrazeDeeplinkHandler.getInstance().gotoUri(appContext, uriAction)
                }
            }
            ClickAction.NONE -> inAppMessageCloser.close(inAppMessage.animateOut)
            else -> inAppMessageCloser.close(false)
        }
    }

    private fun startClearHtmlInAppMessageAssetsThread() {
        BrazeCoroutineScope.launch {
            val inAppMessageActivity = BrazeInAppMessageManager.getInstance().activity
            if (inAppMessageActivity != null) {
                val internalStorageCacheDirectory = getHtmlInAppMessageAssetCacheDirectory(inAppMessageActivity)
                deleteFileOrDirectory(internalStorageCacheDirectory)
            }
        }
    }
}
